### DISCUSSION TITLE
support "const" loop variable in C style for loop

### DISCUSSION THREAD

**Discussion by RokerHRO** [vlang/v#11900] [Ideas]
Documentation says: "Here `i` doesn't need to be declared with mut since it's always going to be mutable by definition."

I'd suggest to change the semantics here:

```
for i := 0; i < 10; i += 2
```

This `i` is "const" in the body of the loop, it is only allowed to be changed by the `i += 2` loop increment instruction.

If the programmer wants to change the loop variable within the loop body freely it **must** be declared with `mut` explicitly:

```
for mut i := 0; i < 10; i += 2
```

Motivation:
Arbitrarily changing the loop variable in the loop body is considered "bad style" and allowing that makes code analysis (by compiler and human readers) much more complex. So it should be not allowed by default and enabled only if necessary.

**Comment by RokerHRO** [comment]
So the issue was closed without comment? So sad. :-(

**Reply by JalonSolov** [reply]
If this was originally an issue, then yes, the issue would've been closed when this was moved to be a discussion.

It was moved to a discussion because, like most feature requests, it needs to have pros and cons hammered out to decide if it is a reasonable change to incorporate, especially when it would be a breaking change (as this one would be).

**Reply by RokerHRO** [reply]
It would only break bad code (=code that changes the loop variable in the for()-loop), but it would make code that _needs_ to change the loop variable more expressive: showing that changing the loop variable is _intentional_, not an accident.

**Reply by JalonSolov** [reply]
It also has the implication that the compiler now has to know "even though this wasn't declared `mut`, it still has to change just to be used in the loop, and that has to be ok".

There are always more things to consider than the one the original requester thought about, hence the need for a discussion.
